casper/src/main/scala/coop/rchain/casper/api/BlockAPI.scala

```diff
@@ -566,20 +572,44 @@ object BlockAPI {
       case None             => none[BlockMessage]
     }
 
-  private def addResponse(
+  private def addResponse[F[_]: Concurrent](
       status: BlockStatus,
-      block: BlockMessage
-  ): ApiErr[DeployServiceResponse] =
+      block: BlockMessage,
+      casper: MultiParentCasper[F],
+      printUnmatchedSends: Boolean
+  ): Effect[F, DeployServiceResponse] =
     status match {
       case _: InvalidBlock =>
-        s"Failure! Invalid block: $status".asLeft
+        s"Failure! Invalid block: $status".asLeft[DeployServiceResponse].pure[F]
       case _: ValidBlock =>
-        val hash = PrettyPrinter.buildString(block.blockHash)
-        DeployServiceResponse(s"Success! Block $hash created and added.").asRight
+        val hash    = PrettyPrinter.buildString(block.blockHash)
+        val deploys = block.body.get.deploys.map(_.deploy.get)
+        val maybeUnmatchedSendsOutputF =
+          if (printUnmatchedSends) prettyPrintUnmatchedSends(casper, deploys)
+          else none[String].pure[F]
+        maybeUnmatchedSendsOutputF >>= (
+            maybeOutput =>
+              DeployServiceResponse(
+                s"Successs! Block $hash created and added.${maybeOutput.map("\n" + _).getOrElse("")}"
+              ).asRight[Error].pure[F]
+          )
       case BlockException(ex) =>
-        s"Error during block processing: $ex".asLeft
+        s"Error during block processing: $ex".asLeft[DeployServiceResponse].pure[F]
       case Processing =>
-        "No action taken since other thread is already processing the block.".asLeft
+        "No action taken since other thread is already processing the block."
+          .asLeft[DeployServiceResponse]
+          .pure[F]
+    }
+
+  private def prettyPrintUnmatchedSends[F[_]: Concurrent](
+      casper: MultiParentCasper[F],
+      deploys: Seq[DeployData]
+  ): F[Option[String]] =
+    casper.getRuntimeManager >>= {
+      case Some(rm) =>
+        rm.withRuntimeLock(runtime => StoragePrinter.prettyPrintUnmatchedSends(deploys, runtime))
```

Comment:
now I see the usage of withRuntimeLock but why do you need to lock the runtime? I'm thinking that the printing of unmatched sends happen after the execution ended so it's not executed concurrently with other runtime operations

Reply:
`StoragePrinter.prettyPrintUnmatchedSends` executes Rholang. Therefore I think we should use `withRuntimeLock`.

Reply:
ok